src/net_processing.cpp

```diff
@@ -2962,13 +2962,14 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman, bool en
         else if (pnode->m_manual_connection)
             LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode->addr.ToString());
         else {
-            pnode->fDisconnect = true;
-            if (pnode->addr.IsLocal())
-                LogPrintf("Warning: not banning local peer %s!\n", pnode->addr.ToString());
-            else
-            {
+            if (pnode->addr.IsLocal()) {
+                // Disconnect but don't ban _this_ local node
+                LogPrintf("Warning: disconnecting but not banning local peer %s!\n", pnode->addr.ToString());
+            } else {
+                // Disconnect and ban all nodes sharing the address
                 connman->Ban(pnode->addr, BanReasonNodeMisbehaving);
             }
```

Comment:
Commit b4ca7e2

Discussed with @dongcarl about not changing behavior in this commit, the conclusion is to call `conman->DisconnectNode(pnode->addr)` here and add `pnode->fDisconnect = true;` above.